Feat: 결제 시스템 구현 - Toss Payments API 연동 및 테스트 Key-in 결제 기능 추가

ZeroColaa opened the pull request:
## #️⃣작업 한줄 요약 :
feat: 결제 시스템 구현 - Toss Payments API 연동 및 테스트 Key-in 결제 기능 추가

## 📝작업 내용> 
주요 변경사항 :

### 1. Payment 도메인 모델 구축
- Payment 엔티티 domain 패키지로 이동 및 구조 개선
- PaymentStatus enum (COMPLETED, REFUNDED)
- 낙관적 잠금(@Version) 및 유니크 제약조건 

### 2. 결제 DTO 클래스 추가
- `CardPaymentTestRequest`: 테스트 Key-in 결제 요청 DTO
- `PaymentResponse`: 결제 응답 DTO (엔티티 -> DTO 변환)
- `RefundRequest`: 환불 요청 DTO

### 3. 리포지토리 계층 구현
- `PaymentRepository` 인터페이스 및 구현체 분리
- 사용자/모임별 결제 내역 조회 기능
- 중복 결제 방지를 위한 검증 메서드

### 4. Toss Payments 연동
- `TossPaymentsClient`: Toss API 호출 클라이언트
- `TossPaymentsConfig`: API 설정 관리
- RestTemplate 설정 및 인증 헤더 자동 추가
- Idempotency Key를 통한 중복 요청 방지

### 5. 결제 서비스 구현
- 테스트 Key-in 결제 처리 (카드번호 직접 입력)
- 무료 모임 결제 자동 처리 (참가비 0원)
- 환불 처리 로직
- 결제 내역 조회 기능

### 6. REST API 엔드포인트
- `POST /api/v1/payments/test/keyin`: 테스트 결제
- `GET /api/v1/payments/meetings/{meetingId}`: 모임별 결제 조회
- `GET /api/v1/payments/users/{userId}`: 사용자별 결제 조회
- `POST /api/v1/payments/{paymentId}/refund`: 환불 처리

##  주요 특징 :
- 테스트 환경에서 Key-in API 사용 가능하도록 검증
- 모임-사용자 간 중복 결제 방지 (유니크 제약조건)
- 모든 응답을 ApiResponse로 표준화


## 예정된 개선 사항 
- 예외 처리는 `BaseException` + `ErrorCode` 기반의 구조로 개선 예정입니다.

ZeroColaa committed:
feat(payment): 결제 API 및 Payment 도메인 기본 구현

ZeroColaa committed:
refactor: Payment 엔티티에서 User, Meeting 연관관계 제거 및 ERD 기반 필드 변경

ZeroColaa committed:
feat: Toss 테스트 결제 연동 및 Payment 도메인 개선

- Payment 엔티티에 orderId 필드 추가 (Toss 주문번호 대응)
- createPaid() 정적 메서드 제거 (key-in 흐름으로 대체)
- Toss API 응답 항목 기반으로 엔티티 필드 보강 및 주석 추가
- 테스트용 키인 결제 API 구현 (/test/keyin)
- 모든 PaymentController 응답 ApiResponse로 통일
- 기존 일반 결제 API 제거, key-in 테스트 흐름만 유지

ZeroColaa committed:
feat: 결제 관련 DTO 클래스 추가

- CardPaymentTestRequest: 테스트 Key-in 결제 요청 DTO
- PaymentResponse: 결제 응답 DTO
- RefundRequest: 환불 요청 DTO

ZeroColaa committed:
feat: Payment 리포지토리 인터페이스와 구현체 추가

- 도메인 계층 인터페이스(PaymentRepository) 정의
- Spring Data JPA 기반 구현(PaymentJpaRepository)
- 구현체(PaymentRepositoryImpl)로 추상화와 분리 적용

ZeroColaa committed:
feat: Toss Payments 연동을 위한 설정 및 클라이언트 추가

- TossPaymentsConfig: Toss API 설정 클래스
- TossPaymentsClient: Toss API 호출 클라이언트
- RestTemplate 설정 및 인증 헤더 자동 추가
- application.yml에 Toss 설정 추가
- @EnableConfigurationProperties 추가

ZeroColaa committed:
feat(payment): 서비스 계층 로직 구현

- PaymentService 인터페이스 정의
- PaymentServiceImpl 구현
- Key-in 테스트 결제 처리 로직 추가
- 결제 환불 처리 로직 추가
- 결제 조회 (사용자/모임별) 기능 구현
- 무료 결제 처리 분기 로직 추가

ZeroColaa committed:
feat: 결제 상태 enum PaymentStatus 추가

ko-dongwon reviewed (approved):
고생하셨습니다!

ZeroColaa committed:
refactor: 결제 예외 처리 개선 및 환불/Key-in 로직 리팩토링

- 도메인 기반 예외 처리 도입 (PaymentException, PaymentErrorCode)
- Toss API 호출 시 try-catch로 감싸고 상세 응답 로그 추가
- Key-in 결제 시 orderId 생성 책임을 메서드 외부로 분리
- 환불 처리 시 상태를 변경한 뒤 결제 데이터 삭제 처리

Mybread2 reviewed (approved):
확인했습니다.

hojin915 commented on src/main/java/com/example/momo/domain/payments/api/PaymentController.java:
```diff
@@ -0,0 +1,58 @@
+  public ResponseEntity<ApiResponse<List<PaymentResponse>>> getPaymentsByUser(
+      @PathVariable Long userId) {
+    List<PaymentResponse> payments = paymentService.getPaymentsByUserId(userId);
+    return ResponseEntity.ok(ApiResponse.success("사용자별 결제 내역 조회가 완료되었습니다.", payments));
```
조회 기능 관리자 측면에서 사용할 것으로 의도하신건가요?
사용자 입장에선 AuthUser 사용해서 본인 내역을 조회할 수 있을 것 같습니다.
그리고 만약 관리자 입장에서 결제 내역을 조회한다면 전체 내역 중에서 MeetingId, UserId로 필터링 하는 방식이 자연스러울 것 같습니다.

ZeroColaa replied:
알겠습니다! 수정해서 merge하겠습니다 피드백 감사합니다

ZeroColaa committed:
feat: 결제·환불 인가 로직 및 요청 유효성 개선

- CardPaymentTestRequest에서 userId 제거 및 DTO 유효성 검증 강화
- Controller에서 인증 사용자 ID만 Service로 전달하도록 변경
- Service에서 결제·환불 요청자의 권한 검증 로직 추가
- PaymentErrorCode에 UNAUTHORIZED_PAYMENT / UNAUTHORIZED_REFUND 추가

ZeroColaa committed:
refactor: Toss Key-in API 호출 시 Idempotency-Key 중복 설정 제거

hojin915 reviewed (approved):
고생하셨습니다

leeuihyun commented on src/main/java/com/example/momo/domain/payments/infra/toss/TossPaymentsClient.java:
```diff
@@ -0,0 +1,103 @@
+  // ==================== Private Helper 메서드 ====================
+
+  /**
+   * Idempotency Key 추가 토스페이먼츠는 15일 동안 동일 키 보관
+   */
+  private void addIdempotencyKey(HttpHeaders headers, String keySeed) {
+    headers.add("Idempotency-Key", keySeed);
+  }
```
토스페이먼츠에서 키를 보관한다는 부분이 Jwt 토큰의 키 만료시간과 같은 걸까요? 키를 보관하게 한 후 `Idempotency-key` 를 헤더에 넣어 보내면 본인임을 확인하는 건가여?

ZeroColaa replied:
Idempotency-Key 는 본인 확인용 인증 수단은 아니고, 동일한 결제 요청이 여러 번 들어왔을 때 중복 결제를 방지하기 위한 키입니다. 토스페이먼츠에서 이 키를 15일 동안 보관해서 같은 키로 요청이 들어오면 기존 요청의 응답을 그대로 반환하게 되어 있습니다.  결제 API의 멱등성 보장 목적입니다! 
 본인 인증은 결제 요청자가 카드 소유자인지를 Toss 서버 측에서 판단하며,  
인증에 실패할 경우 Toss API에서 에러 응답을 반환합니다.
 지금은 test용 api를 사용하고 있기 때문에 결제자 본인 확인을 실제 카드 정보 기반 인증이 아닌 @AuthenticationPrincipal을 통해 로그인 된 사용자 본인이 결제 요청을 수행할 수 있도록 되어있습니다.

leeuihyun replied:
아아 어제 말씀하셨던 그 키가 저 키군요. 확인했습니다

Zyooon reviewed (approved):
고생하셨습니다.

ZeroColaa committed:
merge: dev 브랜치와 병합 (application.yml 충돌 해결)

ZeroColaa committed:
feat: 결제 내역 조회 및 관리자 검색 로직 개선
